extends Control

# æˆ‘æ„Ÿè§‰æˆ‘è¿™é‡Œçš„å†™æ³•æœ‰äº›ä¸‘é™‹, ä½†æ˜¯ä¸çŸ¥é“æ€ä¹ˆæ”¹è¿›
var selected_class_name: String:
	get:
		if class_list.selected_id == []:
			return ""
		var selected_class: ClassInfo = class_info[class_list.selected_id[0]]
		return selected_class.my_class_name

var selected_test_mask: Array[int]:
	get:
		return test_list.button_flag

var selected_test_name: String:
	get:
		if test_list.selected_id == [] or class_list.selected_id == []:
			return ""
		var test_id: int = test_list.selected_id[0]
		var selected_class: ClassInfo = class_info[class_list.selected_id[0]]
		return selected_class.get_test(test_id)

var selected_student_name: String:
	get:
		if student_grid.selected_id == [] or class_list.selected_id == []:
			return ""
		var stu_id: int = student_grid.selected_id[0]
		var selected_class: ClassInfo = class_info[class_list.selected_id[0]]
		return selected_class.get_student(stu_id)

var selected_subject_id: int:
	get:
		return $SubIndication/SubjectOption.selected_subject_id

var export_format: String:
	get:
		return $ExportFormat.export_format

var export_dir_path: String:
	get:
		return $ExportPath.selected_path

var export_file_name: String	# æœ¬æ¥åº”è¯¥å†åšä¸€ä¸ªè®¾ç½®è¾“å‡ºæ–‡ä»¶åç§°çš„åŠŸèƒ½, ç°åœ¨æ‡’å¾—åšäº†

var geo_bio_point_scale: int:
	get:
		return $GeoBioPointScale.point_scale

var api_provider: String:
	get:
		return $ApiProvider.api_provider

var class_info: Array[ClassInfo]

@onready var class_list = $ClassList
@onready var test_list = $TestList
@onready var student_grid = $StudentGrid
@onready var subject_option = $SubIndication/SubjectOption
@onready var export_info_preview = $ExportInfoPreview


func _ready():
	$ExportPath/FileDialog.current_dir = OS.get_system_dir(OS.SYSTEM_DIR_DESKTOP)


func update(new_class_info: Array[ClassInfo]) -> void:
	class_info = new_class_info
	class_list.update(class_info)


func _on_export_pressed() -> void:
	if not examine_data():
		return
	
	var err_code := Interface.export(
		selected_class_name,
		selected_test_mask,
		selected_subject_id,
		selected_student_name,
		export_dir_path,
		'auto',
		export_format,
		geo_bio_point_scale,
		api_provider)
	if err_code == Interface.EXECUTE_SUCCESSFULLY_CODE:
		Constant.warn(self, 'å¯¼å‡ºæˆåŠŸ, å¿«å»çœ‹çœ‹å§')
	else:
		Constant.warn(self, 'å¯¼å‡ºå¤±è´¥, ä¸çŸ¥é“ä¸ºä»€ä¹ˆ, é”™è¯¯ä»£ç [%d]' % [err_code])


func examine_data() -> bool:
	var warn_message := ''
	if selected_class_name == '':
		warn_message += '|éœ€è¦é€‰ä¸€ä¸ªç­çº§|\n'
	if selected_student_name == '':
		Constant.warn(self, '|éœ€è¦é€‰æ‹©ä¸€ä¸ªå­¦ç”Ÿ, ä¸é€‰å­¦ç”Ÿçš„è¯è¯¥è¾“å‡ºè°çš„æ•°æ®å‘¢|\n')

	if selected_test_mask.all(func(i: int): return not i):
		warn_message += 'è‡³å°‘å¾—é€‰æ‹©ä¸€ä¸ªè€ƒè¯•\n'
	
	if export_dir_path == "":
		warn_message += 'è¾“å‡ºè·¯å¾„å‘¢, éœ€è¦è¾“å‡ºæ–‡ä»¶å¤¹çš„ä½ç½®\n'
	
	if selected_subject_id in [Constant.Subject.BIOLOGY, Constant.Subject.GEOGRAPHY]\
		and geo_bio_point_scale == 0:
		warn_message += 'å¦‚æœé€‰çš„æ˜¯åœ°ç†æˆ–è€…ç”Ÿç‰©çš„è¯, éœ€è¦é€‰æ‹©ä¸€ä¸‹åœ°ç†ç”Ÿç‰©çš„æœ€é«˜åˆ†æ•°\n'
		
	if warn_message != "":
		Constant.warn(self, warn_message)
		return false
	
	# å…¶ä»–çš„æ•°æ®éƒ½æœ‰é»˜è®¤å€¼, ä¸ä¼šæ˜¯ç©ºçš„
	return true


func _on_preview_pressed():
	'%s%sæˆç»©åˆ†æ:' % [selected_student_name, Constant.SUBJECT_NAME[selected_subject_id]]
	export_info_preview.text = 'æ²¡åšå®ŒğŸ˜Š, ä¸æƒ³åšäº†'


func _on_class_list_class_selected(id: int):
	test_list.update(class_info[id])
	student_grid.update(class_info[id])
